probe-rs/src/architecture/riscv/communication_interface.rs

```diff
@@ -333,6 +342,28 @@ impl RiscvCommunicationInterface {
         Ok(s)
     }
 
+    /// Select current hart
+    pub fn select_hart(&mut self, hart: u32) -> Result<bool, RiscvError> {
+        if self.enabled_harts & (1 << hart) == 0 {
+            return Ok(false);
```

Comment:
`new` returns `HartUnavailable`, while this function returns `Ok`. For me the semantics here aren't clear - if the hart is not enabled, why return `Ok` and not an error?

Reply:
I see why this looks inconsistent.

The reason is that `Session::new` will call `clear_all_hw_breakpoints` which tries to clear all hw breakpoints on all cores.
Before I made the LP core a separate target this means it would stop with an error when trying to access the LP core if it's not running.

Now I _could_ change things here and in a few other functions to make it nicer when assuming that all cores are available for a target - most probably we would need to change things again when we add support for e.g. P4's second core (if we don't decide to add support for P4's second core via an additional target - which we probably can - but everyone else would need to do that for multicore targets)

Reply:
So the problem is, that we are/were trying to define dissimilar cores in the same list of cores. For now I think it's easiest to just assume that the cores listed in the yamls are the same, especially that you just reverted the change that broke this assumption.

Dissimilar cores will be an issue, for sure, especially if cores appear/disappear after attaching, but that should be a bridge we cross later. For now I believe the cleanest option is to fail if a hart can't be selected, so that the code is consistent. Especially that this solution seems to have weird consequences like reading unintended memory (or at least using the wrong hart for operations) if the hart you want to address has gone away for any reason, if I understand it correctly?

Reply:
The problem is not really dissimilar cores but cores which might not be available depending on whether they are enabled or not but are physically present (and for that reason they will be listed as cores in the yaml)

I guess removing these workarounds here is fine - but we will 100% need something for P4 and others (where one TAP will control 2 or more harts and the some will not always be available)

Reply:
Ok the workaround is removed for now - we'll see what to do when we get to P4's second core